pkg/scheduler/plugins/elasticquota/plugin.go

```diff
@@ -0,0 +1,511 @@
+	if err != nil {
+		klog.ErrorS(err, "Failed to read elasticQuotaSnapshot from cycleState", "elasticQuotaSnapshotKey", SnapshotStateKey)
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+	quotaName := g.getPodAssociateQuotaName(podInfoToAdd.Pod)
+	elasticQuotaInfo := elasticQuotaSnapshotState.elasticQuotaInfos[quotaName]
+
+	if elasticQuotaInfo != nil {
+		if err := elasticQuotaInfo.addPodIfNotPresent(podInfoToAdd.Pod); err != nil {
+			klog.ErrorS(err, "Failed to add Pod to its associated elasticQuota", "pod", klog.KObj(podInfoToAdd.Pod))
+		}
+	}
+
+	return framework.NewStatus(framework.Success, "")
+}
+
+// RemovePod is called by the framework while trying to evaluate the impact
+// of removing podToRemove from the node while scheduling podToSchedule.
+func (g *Plugin) RemovePod(ctx context.Context, state *framework.CycleState, podToSchedule *corev1.Pod,
+	podInfoToRemove *framework.PodInfo, nodeInfo *framework.NodeInfo) *framework.Status {
+	elasticQuotaSnapshotState, err := getElasticQuotaSnapshotState(state)
+	if err != nil {
+		klog.ErrorS(err, "Failed to read elasticQuotaSnapshot from cycleState", "elasticQuotaSnapshotKey", SnapshotStateKey)
+		return framework.NewStatus(framework.Error, err.Error())
+	}
+
+	quotaName := g.getPodAssociateQuotaName(podInfoToRemove.Pod)
+	elasticQuotaInfo := elasticQuotaSnapshotState.elasticQuotaInfos[quotaName]
+
+	if elasticQuotaInfo != nil {
+		if err := elasticQuotaInfo.deletePodIfPresent(podInfoToRemove.Pod); err != nil {
+			klog.Infof("Failed to delete Pod from its associated elasticQuota:%v, pod:%v,err:%v", quotaName, klog.KObj(podInfoToRemove.Pod), err)
+		}
+	}
+
+	return framework.NewStatus(framework.Success, "")
+}
+
+// PostFilter modify the defaultPreemption, only allow pods in the same quota can preempt others.
+func (g *Plugin) PostFilter(ctx context.Context, state *framework.CycleState, pod *corev1.Pod,
+	filteredNodeStatusMap framework.NodeToStatusMap) (*framework.PostFilterResult, *framework.Status) {
+	nnn, status := g.preempt(ctx, state, pod, filteredNodeStatusMap)
+	if !status.IsSuccess() {
+		return nil, status
+	}
+	// This happens when the pod is not eligible for preemption or extenders filtered all candidates.
+	if nnn == "" {
+		return nil, framework.NewStatus(framework.Unschedulable)
+	}
+
+	return &framework.PostFilterResult{NominatedNodeName: nnn}, framework.NewStatus(framework.Success)
+}
+
+func (g *Plugin) Reserve(ctx context.Context, state *framework.CycleState, p *corev1.Pod, nodeName string) *framework.Status {
+	g.Lock()
+	defer g.Unlock()
+
+	quotaName := g.getPodAssociateQuotaName(p)
+	// update groupQuotaManager
+	quotaInfo := g.groupQuotaManager.GetQuotaInfoByName(quotaName)
+	if err := quotaInfo.AssignPod(p); err != nil {
+		klog.Errorf("Failed to add Pod to add pod to its associated quota, error:%v", err.Error())
+		return nil
+	}
+	podRequest, _ := resource.PodRequestsAndLimits(p)
```

Comment:
confuse with addPodIfNotPresent\AssignPod, please merge

Reply:
assignPod only changes the running param of the pod in podCache, the pod must in the podCache before; but addPodIfNotPresent will add the pod to the podCache, the pod must not in the podCache before.